File: 1402-count-square-submatrices-with-all-ones/count-square-submatrices-with-all-ones.ts
function countSquares(matrix: number[][]): number {
    const m = matrix.length, n = matrix[0].length;
    let sum = 0;
    for(let i = 0; i < m; i++) {
        for(let k = 0; k < n; k++) {
            if(matrix[i][k] > 0) {
                sum += matrix[i][k] = Math.min(
                    matrix?.[i - 1]?.[k] ?? 0, 
                    matrix?.[i]?.[k - 1] ?? 0, 
                    matrix?.[i - 1]?.[k - 1] ?? 0
                ) + 1
            }
        }
    }
    return sum
};

File: 1402-count-square-submatrices-with-all-ones/count-square-submatrices-with-all-ones.rs
impl Solution {
    pub fn count_squares(mut matrix: Vec<Vec<i32>>) -> i32 {
        let m = matrix.len();
        let n = matrix[0].len();
        let mut res = 0;
        for i in 0..m {
            for k in 0..n {
                if matrix[i][k] == 1 {
                    let mut min = if let Some (r) = matrix.get(i - 1) { *r.get(k - 1).unwrap_or(&0) } else { 0i32 };
                    let l = if let Some (r) = matrix.get(i ) { *r.get(k - 1).unwrap_or(&0) } else { 0i32 };
                    let d = if let Some (r) = matrix.get(i - 1) { *r.get(k).unwrap_or(&0) } else { 0i32 };
                    min = min.min(l).min(d) + 1;
                    matrix[i][k] = min;
                    res += min;
                }
            }
        }
        res
    }
}